pkg/controller/blackboxtarget/blackboxtarget_controller.go

```diff
@@ -0,0 +1,161 @@
+	err = c.Watch(&source.Kind{Type: &applicationmonitoringv1alpha1.BlackboxTarget{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileBlackboxTarget implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileBlackboxTarget{}
+
+// ReconcileBlackboxTarget reconciles a BlackboxTarget object
+type ReconcileBlackboxTarget struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a BlackboxTarget object and makes changes based on the state read
+// and what is in the BlackboxTarget.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileBlackboxTarget) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling BlackboxTarget")
+
+	// Fetch the BlackboxTarget instance
+	instance := &applicationmonitoringv1alpha1.BlackboxTarget{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	switch instance.Status.Phase {
+	case PhaseFinalizer:
+		return r.setFinalizer(instance)
+	case PhaseDone:
+		reqLogger.Info(fmt.Sprintf("BlackboxTarget phaseDone CR:%s Phase: %v", instance.ObjectMeta.Name, instance.Status.Phase))
+		return r.updatePhase(instance, PhaseReconcileConfig)
+	case PhaseReconcileConfig:
+		return r.reconcileConfig(instance)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// reconcileConfig not sure if these CRs need to be watched for changes, presume I'll do it here
+func (r *ReconcileBlackboxTarget) reconcileConfig(cr *applicationmonitoringv1alpha1.BlackboxTarget) (reconcile.Result, error) {
+	log.Info(fmt.Sprintf("BlackboxTarget reconcileConfig CR:%s Phase: %v", cr.ObjectMeta.Name, cr.Status.Phase))
+
+	bbtList := common.GetBTConfig()
+	crbbtList := cr.Spec.BlackboxTargets
+	crName := cr.ObjectMeta.Name
+
+	// Overwrite the blackboxtargets stored for a particular CR
+	bbtList.BTs[crName] = crbbtList
+
+	// Remove the finalizer so the CR can be deleted
+	if cr.DeletionTimestamp != nil {
```

Comment:
Should this be moved to the top of the function? At the moment it would collect the targets from a deleted CR.

Reply:
Hmm thinking about this more, it gets the upto date targets which are in the AMO CR. I can't move it up any higher really. If you see on line 129, its deleting from the list if the BlackboxTarget CR is being deleted.

Reply:
Hmm do you mean like this?:

`// reconcileConfig not sure if these CRs need to be watched for changes, presume I'll do it here
func (r *ReconcileBlackboxTarget) reconcileConfig(cr *applicationmonitoringv1alpha1.BlackboxTarget) (reconcile.Result, error) {
	log.Info(fmt.Sprintf("BlackboxTarget reconcileConfig CR:%s Phase: %v", cr.ObjectMeta.Name, cr.Status.Phase))

	bbtList := common.GetBTConfig()
	crbbtList := cr.Spec.BlackboxTargets
	crName := cr.ObjectMeta.Name

	// Remove the finalizer so the CR can be deleted
	if cr.DeletionTimestamp != nil {
		// Remove the blackboxtargets assocated with this CR as its being deleted
		delete(bbtList.BTs, crName)
		return reconcile.Result{}, r.removeFinalizer(cr)
	}

	// Overwrite the blackboxtargets stored for a particular CR
	bbtList.BTs[crName] = crbbtList

	return reconcile.Result{}, nil
}`